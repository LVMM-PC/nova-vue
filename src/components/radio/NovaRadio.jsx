import Inventory from '@/utils/inventory';
import Utils from '@/utils/utils';

export default {
  name: 'NovaRadio',
  inject: {
    NovaRadioGroup: {
      default: null
    }
  },
  model: {
    prop: 'checked',
    event: 'update'
  },
  props: {
    prefixedClass: {
      type: String,
      default: `${Inventory.prefix}-radio`
    },
    checked: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    value: {
      type: [Number, Boolean, String],
      default: null
    },
    label: {
      type: [Number, Boolean, String],
      default: ''
    }
  },
  computed: {
    classList() {
      const { isChecked, isDisabled, prefixedClass } = this;

      return [
        prefixedClass,
        {
          'is-checked': isChecked,
          'is-disabled': isDisabled
        }
      ];
    },
    isDisabled() {
      return this.disabled || this.groupDisabled;
    },
    groupValue() {
      return this.NovaRadioGroup?.value;
    },
    groupDisabled() {
      return this.NovaRadioGroup?.disabled;
    },
    isChecked() {
      if (this.NovaRadioGroup) {
        return this.value === this.groupValue;
      } else {
        return this.checked;
      }
    }
  },
  watch: {
    isChecked(newValue, oldValue) {
      if (newValue === oldValue) {
        return;
      }

      this.$emit('update', newValue);
    },
    checked(newValue, oldValue) {
      if (newValue === oldValue) {
        return;
      }

      if (!this.NovaRadioGroup) {
        return;
      }

      if (newValue) {
        this.NovaRadioGroup.setChecked(this.value, false);
      } else {
        const isChecked = this.value === this.NovaRadioGroup.getChecked();
        if (isChecked) {
          this.NovaRadioGroup.setChecked(null, false);
        }
      }
    }
  },
  methods: {
    inGroupClick() {
      this.NovaRadioGroup.setChecked(this.value, true);
    },
    handleRadioClick(e, ...restArgs) {
      if (this.isDisabled) {
        return;
      }

      this.$emit('click', e, ...restArgs);

      const target = e.target;

      const targetIsInput = target instanceof HTMLInputElement;
      const targetIsTextArea = target instanceof HTMLTextAreaElement;

      const $label = this.$refs['label'];
      const familyLink = Utils.getFamilyLink(target, $label);
      const targetInContentEditable = familyLink.find(element => {
        return element.isContentEditable;
      });

      if (!(targetIsInput || targetIsTextArea || targetInContentEditable)) {
        const $input = this.$refs['input'];
        $input?.focus();
        this.setChecked();
      }
    },
    setChecked() {
      if (this.isChecked) {
        return;
      }

      this.$emit('change', true);

      if (this.NovaRadioGroup) {
        this.inGroupClick();
      } else {
        this.$emit('update', true);
      }
    },
    handleKeydown(e) {
      if (this.isDisabled) {
        return;
      }

      switch (e.key) {
        case 'Spacebar': // IE/Edge
        case ' ':
          e.preventDefault();
          this.setChecked();
          break;
        case 'Enter':
          e.preventDefault();
          this.setChecked();
          break;
      }
    }
  },
  render() {
    const {
      $attrs,
      $listeners,
      $slots,
      classList,
      handleKeydown,
      handleRadioClick,
      isDisabled,
      label,
      prefixedClass
    } = this;

    const tabIndex = isDisabled ? -1 : 0;

    const radioProps = {
      class: classList,
      attrs: {
        ...$attrs
      },
      on: {
        ...$listeners,
        click: handleRadioClick
      },
      ref: 'radio'
    };

    const inputProps = {
      class: [`${prefixedClass}-input`],
      attrs: {
        tabindex: tabIndex
      },
      on: {
        keydown: handleKeydown
      },
      ref: 'input'
    };

    const children = $slots.default || label;

    return (
      <div {...radioProps}>
        <div {...inputProps}></div>
        <div class={`${prefixedClass}-label`} ref="label">
          {children}
        </div>
      </div>
    );
  }
};
