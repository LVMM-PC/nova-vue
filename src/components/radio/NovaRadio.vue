<template>
  <div
    ref="radio"
    class="nova-radio"
    v-bind="$attrs"
    v-on="$listeners"
    @click="handleRadioClick"
    :class="classList"
    :tabindex="!isDisabled ? 0 : -1"
  >
    <div class="nova-radio-input">
      <div class="nova-radio-inner"></div>
    </div>
    <div class="nova-radio-label">
      <slot>{{ label }}</slot>
    </div>
  </div>
</template>

<script>
export default {
  name: 'NovaRadio',
  inject: {
    NovaRadioGroup: {
      default: null
    }
  },
  model: {
    prop: 'checked',
    event: 'update'
  },
  props: {
    checked: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    value: {},
    label: {
      type: [Number, Boolean, String],
      default: ''
    }
  },
  computed: {
    classList() {
      return {
        'is-checked': this.isChecked,
        'is-disabled': this.isDisabled
      };
    },
    isDisabled() {
      return this.disabled || this.groupDisabled;
    },
    groupValue() {
      return this.NovaRadioGroup?.value;
    },
    groupDisabled() {
      return this.NovaRadioGroup?.disabled;
    },
    isChecked() {
      if (this.NovaRadioGroup) {
        return this.value === this.groupValue;
      } else {
        return this.checked;
      }
    }
  },
  methods: {
    inGroupClick() {
      this.NovaRadioGroup.setChecked(this.value, true);
    },
    handleRadioClick() {
      if (this.isDisabled) {
        return;
      }

      if (this.isChecked) {
        return;
      }

      this.$emit('change', true);

      if (this.NovaRadioGroup) {
        this.inGroupClick();
      } else {
        this.$emit('update', true);
      }
    }
  },
  watch: {
    isChecked(newValue, oldValue) {
      if (newValue === oldValue) {
        return;
      }

      this.$emit('update', newValue);
    },
    checked(newValue, oldValue) {
      if (newValue === oldValue) {
        return;
      }

      if (!this.NovaRadioGroup) {
        return;
      }

      if (newValue) {
        this.NovaRadioGroup.setChecked(this.value, false);
      } else {
        let isChecked = this.value === this.NovaRadioGroup.getChecked();
        if (isChecked) {
          this.NovaRadioGroup.setChecked(null, false);
        }
      }
    }
  }
};
</script>

<style lang="less">
@import '../../styles/var';

@radio: @{prefixed}-radio;

.@{radio} {
  cursor: pointer;
  line-height: 20px;
  color: #333333;
  display: inline-block;
  vertical-align: top;
  font-size: 12px;
  font-family: @font-family;

  &:focus {
    outline: none;

    &:not(.is-disabled) {
      .@{radio}-input {
        border: 1px solid #ee3388;
      }
    }
  }

  &.is-disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  &:hover:not(.is-disabled) {
    .@{radio}-input {
      border: 1px solid #ee3388;
    }
  }

  &.is-checked {
    .@{radio}-input {
      border: 1px solid #ee3388;
    }

    .@{radio}-inner {
      opacity: 1;
      transform: scale(1);
    }
  }
}

.@{radio}-input {
  display: inline-block;
  vertical-align: top;
  width: 14px;
  height: 14px;
  margin-top: 3px;
  margin-right: 4px;
  border: 1px solid #cccccc;
  box-sizing: border-box;
  border-radius: 50%;
}

.@{radio}-inner {
  opacity: 0;
  width: 8px;
  height: 8px;
  transform: scale(0);
  transition: transform @fast-motion @ease-in-out-circular;
  background-color: #ee3388;
  border-radius: 50%;
  margin: 2px;
}

.@{radio}-label {
  display: inline-block;
  vertical-align: top;
}
</style>
