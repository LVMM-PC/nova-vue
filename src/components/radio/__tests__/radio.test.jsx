import mountTest from '../../../../tests/shared/mountTest';
import NovaRadio from '@/components/radio/NovaRadio';
import { mount } from '@vue/test-utils';

describe('NovaRadio.vue', () => {
  mountTest(NovaRadio);

  it('should render correctly', () => {
    const wrapper = mount({
      render() {
        return <NovaRadio class="customized" />;
      }
    });
    expect(wrapper).toMatchSnapshot();
  });
});
