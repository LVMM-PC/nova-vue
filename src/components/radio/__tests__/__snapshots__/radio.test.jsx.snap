// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NovaRadio should render correctly 1`] = `
<div class="nova-radio customized">
  <div tabindex="0" class="nova-radio-input"></div>
  <div class="nova-radio-label"></div>
</div>
`;
